packages/e2e-tests/test-utils/event-proxy-server.ts

```diff
@@ -0,0 +1,220 @@
+    proxyServer.listen(options.port, () => {
+      resolve();
+    });
+  });
+
+  const eventCallbackServer = http.createServer((eventCallbackRequest, eventCallbackResponse) => {
+    eventCallbackResponse.statusCode = 200;
+    eventCallbackResponse.setHeader('connection', 'keep-alive');
+
+    const callbackListener = (data: string): void => {
+      eventCallbackResponse.write(data.concat('\n'), 'utf8');
+    };
+
+    eventCallbackListeners.add(callbackListener);
+
+    eventCallbackRequest.on('close', () => {
+      eventCallbackListeners.delete(callbackListener);
+    });
+
+    eventCallbackRequest.on('error', () => {
+      eventCallbackListeners.delete(callbackListener);
+    });
+  });
+
+  const eventCallbackServerStartupPromise = new Promise<void>(resolve => {
+    const listener = eventCallbackServer.listen(0, () => {
+      const port = String((listener.address() as AddressInfo).port);
+      const tmpFileWithPort = path.join(os.tmpdir(), `event-proxy-server-${options.proxyServerName}`);
+      fs.writeFileSync(tmpFileWithPort, port, { encoding: 'utf8' });
+      resolve();
+    });
+  });
+
+  await eventCallbackServerStartupPromise;
+  await proxyServerStartupPromise;
+  return;
+}
+
+export function waitForRequest(
+  proxyServerName: string,
+  callback: (eventData: SentryRequestCallbackData) => boolean,
+): Promise<SentryRequestCallbackData> {
+  const tmpFileWithPort = path.join(os.tmpdir(), `event-proxy-server-${proxyServerName}`);
+  const eventCallbackServerPort = fs.readFileSync(tmpFileWithPort, 'utf8');
+
+  return new Promise<SentryRequestCallbackData>((resolve, reject) => {
+    const request = http.request(`http://localhost:${eventCallbackServerPort}/`, {}, response => {
+      let eventContents = '';
+
+      response.on('error', err => {
+        reject(err);
+      });
+
+      response.on('data', (chunk: Buffer) => {
+        const chunkString = chunk.toString('utf8');
+        chunkString.split('').forEach(char => {
+          if (char === '\n') {
+            const eventCallbackData: SentryRequestCallbackData = JSON.parse(
+              Buffer.from(eventContents, 'base64').toString('utf8'),
+            );
+            if (callback(eventCallbackData)) {
+              response.destroy();
+              resolve(eventCallbackData);
+            }
+            eventContents = '';
+          } else {
+            eventContents = eventContents.concat(char);
+          }
+        });
+      });
+    });
+
+    request.end();
+  });
+}
+
+export function waitForEnvelopeItem(
+  proxyServerName: string,
+  callback: (envelopeItem: EnvelopeItem) => boolean,
+): Promise<EnvelopeItem> {
+  return new Promise((resolve, reject) => {
+    waitForRequest(proxyServerName, eventData => {
+      const envelopeItems = eventData.envelope[1];
+      for (const envelopeItem of envelopeItems) {
+        if (callback(envelopeItem)) {
+          resolve(envelopeItem);
+          return true;
```

Comment:
m: Why are we returning from the promise? Ditto for the other `waitForX` methods.

Reply:
We're not returning from the promise, we're returning from the callback of `waitForRequest`. This is to stop listening for events. Do you know a better way to do this?

Reply:
yup you're right, the nesting just confused me.

I think it's fine, just takes a bit to get - this is the best way to share state this functional way.

The alternative would be to have this all be in a class mutating some internal state, but that is prob less clean.